Update androidx-compose-bom dependency to use '2024.09.01'

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Please do not merge before target is `trunk`. It is currently targeting a different branch that is also about dependency update #13046. It seems safest to get that dependency update merged first, and then this one after, to avoid any trouble.

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR updates the version used on `androidx-compose-bom` to '2024.09.00'.

**Why '2024.09.00'?** Originally, I wanted to add these dependencies:
```
androidx.compose.material3.adaptive:adaptive | 1.0.0
androidx.compose.material3.adaptive:adaptive-layout | 1.0.0
androidx.compose.material3.adaptive:adaptive-navigation | 1.0.0
```

and these were made available since '2024.09.00'. To confirm [you can go here](https://developer.android.com/develop/ui/compose/bom/bom-mapping) and change the version in the dropdown to `2024.09.00` and compare it to the version before it.

Update: this is now updated to **2024.09.01** instead. The reason is [from the previous PR:](https://github.com/woocommerce/woocommerce-android/pull/13046#issuecomment-2519971988):

> Lifecycle Runtime Compose now depends on [Compose Runtime 1.7.1](https://developer.android.com/jetpack/androidx/releases/compose-runtime#1.7.1)

I had thought about updating to a newer version, but as you can see from the commits, this alone already introduces quite a bit of required changes, so I'm a bit wary about going even further.

I apologize for the size of the PR. Most of them are repeat changes related to these:

| No. | Updates | Notes |
|-|-|-|
| 1. | deprecated `rememberRipple()` usage | Trivial syntax change |
| 2. | deprecated `RippleTheme` and `LocalRippleTheme` usage. | More info in Testing Information section |
| 3. | deprecated `LocalLifecycleOwner` usage. | Trivial syntax change |
| 4. | deprecated `animateItemPlacement()` usage. | Trivial syntax change |
| 5. | deprecated `KeyboardOptions` constructor. | Trivial syntax change |
| 6. | deprecated `updateTransition` usage. | Trivial syntax change |
| 7. |  deprecated `ClickableText` usage. | More info in Testing Information section |
| 8. | deprecated `LocalTextInputService` usage | More info in Testing Information section |

These are all warnings-level change, not error. But since build.gradle is set to treat `allWarningsAsErrors`, they need to be updated.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Make sure the app builds with no errors
- Smoke test the main flows P91TBi-bVe-p2

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

**On ClickableText:**
I've tested them and can confirm that they are working fine, but could use a double-check.  Please also double-check the code changes on commit 5cc7b7a1251a16cde1c6f07ad3f48227ada98196 and da33b40c78441c3ee9287e67f469577440c580ca to emsure I'm not copy/pasting the wrong string resource/onclick functions.

The affected screens are:
- JetpackConsent
- AddProductWithAIBottomSheet
- BlazeCampaignPaymentSummaryScreen
- WidgetError
- AccountMismatchErrorScreen
- PaymentsPreSetupScreen
- WooPaymentsSetupInstructionsScreen
- DomainDashboardScreen

**On RippleTheme:**
The `RippleTheme` replacement (in this commit https://github.com/woocommerce/woocommerce-android/pull/13060/commits/145a2170e2277629b9d4736a3233389b4810672c) now does not have a custom Alpha compared to the previous one. This is because now [RippleAlpha](https://developer.android.com/reference/kotlin/androidx/compose/material/ripple/RippleAlpha) has different constructor while `defaultRippleAlpha` is deprecated, and I can't figure out how to replicate it. With an eye test I don't see a difference between the ripple effect in `trunk` and this branch (I test using the "Share" button in Product Detail's share bottom sheet), but suggestions welcome if updating it further is needed.

**On the focus fix** (commits a3716c6bd2bdd5b26204ec284b569d7c2431ce9e and efcef5c20177e6cf22e7b7efaa7f1494bb0753ef)
This is related to Aztec usage in Custom Fields Editor and has been discussed in Slack p1733297997528939-slack-C03L1NF1EA3 . @hichamboushaba and I have tested this fix so it's okay to not be tested again. The **tl;dr** is that the original code contained a deprecated use of `LocalTextInputService`, but we're lucky that it's addressing an issue [that's already fixed](https://issuetracker.google.com/issues/318530776), so we could clean up the code and avoid the deprecation warning at the same time. 

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
I've smoke tested this on the flow P91TBi-bVe-p2
Also checked the affected clickable texts, and RippleTheme effect, and the Aztec keyboard related fix above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
n/a

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Update compose-bom version to use 2024.09.00

hafizrahman committed:
Update ripple() usage.

hafizrahman committed:
Update deprecated `RippleTheme` and `LocalRippleTheme` usage.

hafizrahman committed:
Replace deprecated usage of LocalLifecycleOwner

Warning was: 'val LocalLifecycleOwner: ProvidableCompositionLocal<LifecycleOwner>' is deprecated. Moved to lifecycle-runtime-compose library in androidx.lifecycle.compose package.

hafizrahman committed:
Replace deprecated `animateItemPlacement()` usage

Warning was: 'fun Modifier.animateItemPlacement(animationSpec: FiniteAnimationSpec<IntOffset> = ...): Modifier' is deprecated. Use Modifier.animateItem() instead.

hafizrahman committed:
Replace deprecated KeyboardOptions constructor.

Warning was: 'constructor(capitalization: KeyboardCapitalization = ..., autoCorrect: Boolean, keyboardType: KeyboardType = ..., imeAction: ImeAction = ..., platformImeOptions: PlatformImeOptions? = ..., showKeyboardOnFocus: Boolean? = ..., hintLocales: LocaleList? = ...): KeyboardOptions' is deprecated. Please use the new constructor that takes optional autoCorrectEnabled parameter.

hafizrahman committed:
Replace deprecated use of `updateTransition`

Warning was: 'fun <T> updateTransition(transitionState: MutableTransitionState<T>, label: String? = ...): Transition<T>' is deprecated. Use rememberTransition() instead.

hafizrahman committed:
Add missing replacement for deprecated usage of LocalLifecycleOwner

hichamboushaba committed:
An attempt to update annotatedStringRes to work with LinkAnnotation

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
More replacement for deprecated uses of ClickableText

hafizrahman committed:
Various automated detekt fixes.

hafizrahman committed:
Update ripple color usage.

hafizrahman committed:
Update the way clickable link is built in CustomFields screen.

hafizrahman committed:
Attempt to remove focus fix as the bug have been fixed

https://issuetracker.google.com/issues/318530776 is where it says it is now fixed. This also conveniently updates the deprecated usage of LocalTextInputService

hafizrahman commented on gradle/libs.versions.toml:
```diff
@@ -9,7 +9,7 @@
 androidx-camera = '1.2.3'
 androidx-cardview = '1.0.0'
-androidx-compose-bom = '2024.04.00'
+androidx-compose-bom = '2024.09.00'
```
I have explained in the original post above why I chose 2024.09.00

hafizrahman committed:
Merge branch 'hack-week/update-dep-androidx-lifecycle' into hack-week/update-dep-compose-bom

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/customfields/CustomOrderFieldsScreen.kt

hichamboushaba committed:
Bring back the scroll logic when focus changes

hafizrahman committed:
Update release notes

hafizrahman commented:
@JorgeMucientes as mentioned in https://github.com/woocommerce/woocommerce-android/pull/13046#issuecomment-2522173308 , I'll bump this to 2024.09.01 instead of 2024.09.00, so I'm setting this back to draft for now. Will let you know once it's done.

hafizrahman committed:
Bump compose-bom to 2024.09.01

hafizrahman commented:
@JorgeMucientes Not seeing any build issue after upgrading to 2024.09.01, so setting this to ready for review again 🙏🏼

JorgeMucientes reviewed (approved):
Great job @hafizrahman. Thanks for proactively tackling this update. 

Code changes look good and I smoke tested the app and everything worked well. 

About the Ripple effect
> With an eye test I don't see a difference between the ripple effect in trunk and this branch (I test using the "Share" button in Product Detail's share bottom sheet), but suggestions welcome if updating it further is needed.

I think the difference is noticeable only in dark mode, where the ripple effect is now applying `<color name="color_primary_variant">@color/woo_purple_50</color>` which is a darker color than the `color_primary` set. I'm not sure how it worked before, but visually you could see the button got brighter upon tapping on it. Now it gets slightly darker. Barely noticeable. This is a minor thing, so definitely not blocking and can be addressed in a different PR. Maybe we can refine the color used in dark mode to have a darker ripple similar to the one the FAB button has.  

PS: I tested updating the BOM to the latest version during my hackweek (using this brach as base) and I didn't have to apply any code changes to get the app to compile. So hopefully the next BOM update will be painless, if we don't pospone updating to much :)

hafizrahman commented:
@JorgeMucientes Thanks for checking out the Ripple situation. We're also likely updating colors soon, so probably a good idea to not spend too much time on this now.
